src/Formality__FormAsyncOnChange.re

```diff
@@ -100,10 +102,15 @@ module Make = (Form: Config) => {
         Validation.validationResult(Form.message),
       )
     | Submit
-    | Reset
-    | HandleSubmissionError;
+    | SetSubmittedStatus(option(Form.state))
+    | SetSubmissionFailedStatus(
+        list((Form.field, Form.message)),
+        option(Form.message),
+      )
+    | Reset;
   type interface = {
     state: Form.state,
+    status: FormStatus.t(Form.field, Form.message),
```

Comment:
Why does status need a field?

Reply:
B/c of this tag: `SubmissionFailed(list(('field, 'message)), option('message))`

Reply:
So you're keeping the client validation state separate from the server validation state?

Reply:
Yep, I kept them separate b/c of the following case: user submits form and server responds w/ errors. If I replace client errors w/ errors from server, then on the next change in failed field server's error will be wiped out by client's result. If client result is `Valid`, this means that we signal to user that problem is solved which is probably not true. We can tell it for sure only after re-submission.

Usually, I try to validate as much as possible on the client (using sync & async validations) . So in this context, errors from server after submission might happen in case of different regexes or values (like min/max etc). Basically, due to bug in app code. And in this case I would let user know where the problem is and keep this reason on the screen until next submission attempt.

Reply:
I think another corner case of a server error happening besides an app bug is if you are creating a unique key (like a username).  You can validate uniqueness through sync/async server calls as the user types but a race condition exists... only after submitting when the database server attempts the final INSERT in the transaction do we know it was truly successful.  If a collision occurs the database error can be returned and be kept on the screen until the next submission attempt.

Reply:
@rauanmayemir I had a thought to store these errors under different key in internal state and then merge them into `results` map before passing to children (e.g. if client is invalid -> show client error, if client is valid but server has error -> show server error). But the nature of these errors is too different and it will be wrong UX wise to handle them in UI the same way. Consider the case: if client error is shown, it means user must do something before attempting to re-submit the form (server won’t be touched until all client validations are pass). But when the client is satisfied, we switch client error w/ server error that looks exactly the same as client one in UI, so it’s totally unclear that at this point user can re-submit the form b/c he prolly fixed the issue.